rpc: simplify getaddressinfo labels, deprecate previous behavior

jonatack opened the pull request:
This PR builds on #17283 (now merged) and is followed by #17585.

It modifies the value returned by rpc getaddressinfo `labels` to an array of label name strings and deprecates the previous behavior of returning an array of JSON hash structures containing label `name` and address `purpose` key/value pairs.

before
```
  "labels": [
    {
      "name": "DOUBLE SPEND",
      "purpose": "receive"
    }
```
after
```
  "labels": [
    "DOUBLE SPEND"
  ]
```

The deprecated behavior can be re-enabled by starting bitcoind with `-deprecatedrpc=labelspurpose`.

For context, see:
- https://github.com/bitcoin/bitcoin/pull/17283#issuecomment-554458001
- http://www.erisian.com.au/bitcoin-core-dev/log-2019-12-13.html#l-425 (lines 425-427)
- http://www.erisian.com.au/bitcoin-core-dev/log-2019-11-22.html#l-622

Reviewers: This PR may be tested manually by building, then running bitcoind with and without the `-deprecatedrpc=labelspurpose` flag while verifying the rpc getaddressinfo help text and `labels` output.

Next steps: deprecate the rpc getaddressinfo `label` field (EDIT: done in #17585) and add support for multiple labels per address. This PR will unblock those.

jonatack committed:
rpc: simplify getaddressinfo labels, deprecate previous behavior

- change the value returned in the RPC getaddressinfo `labels` field to an array
  of label name strings

- deprecate the previous behavior of returning a JSON hash structure containing
  label `name` and address `purpose` key/value pairs

- update the relevant tests

jonatack committed:
test: getaddressinfo labels purpose deprecation test

jonatack committed:
doc: update release notes

Update the release notes regarding the change in rpc getaddressinfo `labels`.

jnewbery commented on test/functional/test_runner.py:
```diff
@@ -211,6 +211,7 @@
     'p2p_permissions.py',
     'feature_blocksdir.py',
     'feature_config_args.py',
+    'rpc_getaddressinfo_labels_purpose_deprecation.py',
```
I think it's better to just add the test to the `rpc_deprecated.py` test file.

jonatack replied:
> I think it's better to just add the test to the `rpc_deprecated.py` test file.

I started to, then quickly saw that this would completely change that file (here we are not testing a removed RPC but a change in a returned value), and restoring this test file after 0.20 is released would be more complicated than deleting a test file.

jnewbery reviewed (commented):
Concept ACK. A few comments inline.

jnewbery commented:
I also think you need to squash _pc: simplify getaddressinfo labels, deprecate previous behavior_ and _test: update getaddressinfo labels tests_ commits. I think the intermediate commit with the RPC changes but not the test changes breaks git bisect.

jonatack committed:
rpc: incorporate review feedback from PR 17283

- (reverted after follow-on review by maintainers: provide a valid address in getaddressinfo RPCExample)
- remove unneeded code comments

jnewbery commented:
rebase please!

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3709,60 +3709,67 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
             "    [\n"
+            "      \"label name\" (string) The label name. Defaults to \"\". Equivalent to the label field above.\n"
+            "      ,...\n"
+            "    DEPRECATED:\n"
```
Done.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3802,15 +3825,24 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+        } else {
+            labels.push_back(mi->second.name);
+        }
+        ret.pushKV("labels", std::move(labels));
```
I didn't see an actual change in behavior by putting this line inside the conditional, but I'll go with less change. Done.

jonatack commented:
> I also think you need to squash _pc: simplify getaddressinfo labels, deprecate previous behavior_ and _test: update getaddressinfo labels tests_ commits. I think the intermediate commit with the RPC changes but not the test changes breaks git bisect.

Squashed.

Thanks @jnewbery for reviewing!  I took all your all suggestions except for merging the deprecation test into rpc_deprecated.py (as per my comment above), added a commit incorporating your review feedback from #17283, and rebased. I also beefed up the deprecation test a bit and moved the release note to its own file rather than modifying release-notes.md directly.

jnewbery commented on test/functional/rpc_getaddressinfo_labels_purpose_deprecation.py:
```diff
@@ -0,0 +1,52 @@
+    def run_test(self):
+        """Test getaddressinfo labels with and without -deprecatedrpc flag."""
+        self.test_labels_with_deprecatedrpc_flag()
+        self.test_labels_without_deprecatedrpc_flag()
```
Nit: I think this would be clearer as just a single `run_test()` function, since these helper functions are only used in one place and they're going to be deleted in v0.21 anyway.

jonatack replied:
Done.

jnewbery reviewed (commented):
utACK 583be1e156343f1e7c734fcb9d04db03e96f4cac

One style nit in the test. Feel free to take or ignore.

jonatack commented:
Thanks for reviewing @jnewbery. Addressed the style nit and rebased.

jnewbery commented:
ACK 2e06da5d62e5b3563001d73b62e90a5a61b47993

jonatack commented:
Rebased.

jnewbery commented:
ACK 55dfbc80c37296d6cab736ab267d887840b22367

Only change is rebasing on master.

instagibbs commented:
What's the best argument in favor of the "purpose" field? Aside from it being "strange" I didn't see any discussion really on the links? I'd like to know why it was there before removing it.

jnewbery commented:
"purpose" was only exposed in this RPC in a recent release, merged in https://github.com/bitcoin/bitcoin/pull/12892 (originally from the commit here: https://github.com/bitcoin/bitcoin/pull/7729/files#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR3828).

I expect no-one is using it.

instagibbs commented:
@jnewbery Yes I read that particular fact in the OP. Maybe just saying "why it's weird" is enough, I just have no preconception here.

jonatack commented:
It's worthwhile to ask. I think the [irc comments yesterday](http://www.erisian.com.au/bitcoin-core-dev/log-2019-12-13.html#l-425) (lines 426-426) by phantomcircuit were helpful; added a link to them in the PR description.

jonatack commented:
It's worthwhile to ask. I think the [irc comments yesterday](http://www.erisian.com.au/bitcoin-core-dev/log-2019-12-13.html#l-425) (lines 425-427) by phantomcircuit are helpful and added a link to them in the PR description.

fjahr commented:
ACK 55dfbc80c37296d6cab736ab267d887840b22367

Reviewed code, ran tests and tested behavior manually.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3708,6 +3708,8 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
         return NullUniValue;
     }
 
+    const std::string example_address = "\"bc1qz0wjhhvtlzqec26ftsrwpkwyevjdm7apwctfvx\"";
```
I liked that the address was invalid, so it is harder to accidentally send funds to it

jonatack replied:
I'll change it back *and* add a mention in the developer notes rpc section then :smile:

jonatack replied:
Though I agree with @jnewbery that it's nicer when the rpc examples actually work when copy-pasted. It's a trade-off, to be sure.

MarcoFalke replied:
It *does* work. It is an example for an invalid address ;)

jonatack replied:
Done!

jnewbery replied:
> so it is harder to accidentally send funds to it

Huh? This is the help text for `getaddressinfo()`. It doesn't send funds.

This help text originally showed a valid address, then was changed to invalid in 17283 (https://github.com/bitcoin/bitcoin/pull/17283#discussion_r350784833), and has now been changed to valid again and invalid again. I don't think it matters that much, but can we just pick one and stick to it?

Also agree that the developer notes should be updated to stop bikeshedding around this.

jonatack commented:
Changed back to using an invalid address for the RPCExamples and rebased.

jnewbery commented:
Code review ACK 4752a2a37f63063656e5e0ef02bb861c605c2931

luke-jr commented on doc/release-notes-17578.md:
```diff
@@ -0,0 +1,8 @@
+--------------------------
+
+- The `getaddressinfo` RPC `labels` field now returns an array of label name
+  strings. Previously, it returned an array of JSON hashes containing `name` and
```
JSON has Objects, not hashes.

jonatack replied:
done

jonatack commented:
Doc-only changes after @luke-jr's review:
```diff
doc/release-notes-17578.md
-  strings. Previously, it returned an array of JSON hashes containing `name` and
+  strings. Previously, it returned an array of JSON objects containing `name` and
```
```diff
src/wallet/rpcwallet.cpp
-    // value of the name key/value pair in the labels hash array below.
+    // value of the name key/value pair in the labels array below.
```

@luke-jr, @fjahr, @jnewbery, @MarcoFalke, @weerayut3, @instagibbs, would you mind ACKing (or re-ACKing)?

jnewbery commented:
reACK 8925df8

promag reviewed (commented):
I thought the only change in tests should be to enable `labelspurpose`? Then these test changes would be done when the deprecated behavior is removed.

jonatack commented:
@promag The `labelspurpose` deprecation itself is tested in 8bb405bbadf11391ccba7b334b4cfe66dc85b390. That test can be deleted after the 0.21 release.

promag commented on doc/release-notes-17578.md:
```diff
@@ -0,0 +1,8 @@
+- The `getaddressinfo` RPC `labels` field now returns an array of label name
+  strings. Previously, it returned an array of JSON objects containing `name` and
+  `purpose` key/value pairs, which is now deprecated and will be removed in
+  0.21. To re-enable the previous behavior, launch bitcoind with
```
nit, just `launch with -deprecatedrpc...`.

jonatack replied:
Thanks; will remove "bitcoind" here if need to retouch this PR, otherwise will change it in one of the follow-on PRs like #17585 or in the wiki.

promag replied:
Sure, just to clarify, I'd drop `bitcoind` because the argument also applies to `bitcoin-qt`.

jonatack replied:
Yes, agreed.

jonatack replied:
Done in #17585

promag reviewed (commented):
Code review ACK 8925df86c4df16b1070343fef8e4d238f3cc3bd1.

meshcollider reviewed (commented):
Code review ACK 8925df86c4df16b1070343fef8e4d238f3cc3bd1

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3819,15 +3820,22 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+    // stable if we allow multiple labels to be associated with an address in
+    // the future.
+    //
+    // DEPRECATED: The previous behavior of returning an array containing a JSON
```
thanks; done in #17585
